### PR TITLE
Implement configuration of webserver listening address

### LINKED ISSUE #2774: Adding the opportunity to select Falco webserver port in configuration file
**Motivation**

When I deploy Falco on our Linux infrastructure, I see that it starts listening on port: 8765/TCP on all host IPs (it also includes the external IP if there is one).
And we can get the Falco web server port available on the entire Internet on our servers if we do not prepare firewall rules. Perhaps now there are only two available endpoints, access to which will not give a potential attacker the opportunity to hack us, but we at least reveal to the world security tools we use, which is not very good.
I think I found responsible [line](https://github.com/falcosecurity/falco/blob/f7c628f6237d5a6f7a33e2ed053ced70386193e3/userspace/falco/webserver.cpp#L83) of code.

**Feature**

To add an option in **falco.yaml** configuration file to select IP address for Falco web server between 0.0.0.0 and 127.0.0.1 (with default 0.0.0.0)

**Alternative**

Give a user an opportunity to select an arbitrary IP address for the Falco web server and validate it through regular expression or compare with available network interfaces on host.

### PR DESCRIPTION
**Pull request by sgaist**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

> /kind release

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

/area engine

> /area tests

> /area proposals

> /area CI

**What this PR does / why we need it**:

Currently the webserver is listening on the hard coded 0.0.0.0 address.
This patch keeps this default but allows the administrator to change it through the `falco.yaml` file.

**Which issue(s) this PR fixes**:

Fixes #2774

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
feat(userspace/falco): falco administrators can now configure the address on which the webserver listen using the new listen_address field in the webserver block of the `falco.yaml` file.
```

### CODE REVIEW

File: falco.yaml
Comment (incertum) [inline falco.yaml:495]: Should we explicitly mention that only TCP/IPv4 is supported? Compare to grpc bind_address implementation (which btw perhaps could be cleaned up as well by someone in the future).

**Review by incertum — COMMENTED** [review]
Thank you @sgaist!

Totally makes sense to me and implementation LGTM.
Since I am not using this functionality myself, let's have another maintainer take a closer look.

CC @LucaGuerra @FedeDP thanks!

File: userspace/falco/configuration.cpp
Comment (FedeDP) [inline userspace/falco/configuration.cpp:36]: Can we add some tests around this? Eg here: https://github.com/falcosecurity/falco/tree/master/unit_tests/falco/app/actions

**Review by FedeDP — COMMENTED** [review]
LGTM, i  really like it, thank you!
Left only a comment!

File: userspace/falco/configuration.cpp
Comment (incertum) [reply userspace/falco/configuration.cpp:36]: +1 on some unit tests.

We more recently refactored a few things to make unit testing easier, however you will notice we still have a lot of work ahead, especially in terms of coverage.

Therefore, as minimum test requirement testing the regex only seems fair, for example checkout the prometheus interval test https://github.com/falcosecurity/falco/blob/1609ee85596cd4e57c15423e17a54fe1eb329092/unit_tests/engine/test_falco_utils.cpp#L41

We also have a newer repo here https://github.com/falcosecurity/testing.

As said additional tests and more future contribution in this regard would be appreciated, but only a regex test should be the requirement to get this PR merged.

Thanks in advance!

File: falco.yaml
Comment (sgaist) [reply falco.yaml:495]: Do you mean TCP/IP VS unix socket ?

File: userspace/falco/configuration.cpp
Comment (incertum) [reply userspace/falco/configuration.cpp:37]: Feel free to implement the best test. Happy to provide more feedback once it is up. Just wanted to communicate the absolute minimum test requirement beforehand as you will see that currently not all configurations are tested in this repo's unit test suite :)

Comment (incertum) [reply userspace/falco/configuration.cpp:37]: EDIT: I see it's up, one sec!

File: falco.yaml
Comment (incertum) [reply falco.yaml:495]: And Ipv4 only.

File: unit_tests/falco/test_configuration.cpp
Comment (incertum) [inline unit_tests/falco/test_configuration.cpp:178]: Since we are now accepting user input, lets also check for some typos such as accidental white spaces and such for completeness? WDYT?

File: falco.yaml
Comment (sgaist) [reply falco.yaml:495]: On the TCP side, shouldn't IPV6 also work with httplib ?

File: unit_tests/falco/test_configuration.cpp
Comment (sgaist) [reply unit_tests/falco/test_configuration.cpp:178]: Sure thing

File: unit_tests/falco/test_configuration.cpp
Comment (incertum) [inline unit_tests/falco/test_configuration.cpp:207]: :smile:

File: falco.yaml
Comment (incertum) [reply falco.yaml:496]: You wanna try? If we easily can, we should. Perhaps we just need the ipv6 regex parser then?

Comment (FedeDP) [reply falco.yaml:496]: This LGTM; re ipv6, i think it is just a mattter of updating the regex (and tests) as Melissa said.
We can do that in a follow up PR anyway!

Comment (sgaist) [reply falco.yaml:496]: Sorry for the late reply, I did a small test to see if the webserver starts when listening to an IPV6 address other than `::` (which is working and thus already a good sign) and it does for pure IPV6 addresses. From the looks of it, literal IPV6 addresses are not supported. I currently don't know whether IPV4 mapped address work either.

Related to that topic, we might want to discuss things a bit further with regard to the IPV6 support design. Do we just reuse the same setting but with a regex that can validate both IPV4 and IPV6 values ? Or should there be a separation between both through the use of `listen_address` and `listen_ip6_address` ?

Comment (Andreagit97) [reply falco.yaml:496]: > Do we just reuse the same setting but with a regex that can validate both IPV4 and IPV6 values ?

I would say yes. I would reuse the same config allowing users to specify either an ipv4 or an ipv6

File: userspace/falco/configuration.cpp
Comment (incertum) [inline userspace/falco/configuration.cpp:37]: 😅

**Review by incertum — COMMENTED** [review]
LGTM, will approve after you re-push the last commit with signed off, ty!

File: userspace/falco/configuration.cpp
Comment (sgaist) [reply userspace/falco/configuration.cpp:37]: "savage" might be a good description 😁

**Comment by sgaist** [comment]
> LGTM, will approve after you re-push the last commit with signed off, ty!

sign-off added and re-based as well :-)

**Review by incertum — APPROVED** [review]
/approve

@sgaist 🎉  thanks so much for this contributions, hopefully we see more in the future 🙃 !

File: userspace/falco/configuration.cpp
Comment (FedeDP) [reply userspace/falco/configuration.cpp:37]: Why did you share my password manager passphrase here?? :rofl:

**Review by FedeDP — APPROVED** [review]
/approve

Great job, thank you very much!

**Pull request merged**
